src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Telemetry/VsTelemetryService.cs

```diff
@@ -0,0 +1,238 @@
+                description: description,
+                exceptionObject: exception,
+                gatherEventDetails: callback);
+        }
+
+        /// <summary>
+        /// Posts an event with a single property.
+        /// </summary>
+        /// <param name="eventName">Name of the event</param>
+        /// <param name="propertyName">Name of the property.</param>
+        /// <param name="propertyValue">Value of the property.</param>
+        /// <param name="unconfiguredProject">Correlated project asset for this event.</param>
+        public void PostProperty(string eventName, string propertyName, string propertyValue, UnconfiguredProject unconfiguredProject = null)
+        {
+            TelemetryEvent telemetryEvent = new TelemetryEvent(EventPrefix + eventName.ToLower());
+            telemetryEvent.Properties.Add(BuildPropertyName(eventName, propertyName), propertyValue);
+            TryCorrelateProject(unconfiguredProject, telemetryEvent);
+            TelemetryService.DefaultSession.PostEvent(telemetryEvent);
+        }
+
+        /// <summary>
+        /// Posts an event with multiple properties.
+        /// </summary>
+        /// <param name="eventName">Name of the event.</param>
+        /// <param name="properties">Collection of property name and values.</param>
+        /// <param name="unconfiguredProject">Correlated project asset for this event.</param>
+        public void PostProperties(string eventName, IEnumerable<(string propertyName, string propertyValue)> properties, UnconfiguredProject unconfiguredProject = null)
+        {
+            TelemetryEvent telemetryEvent = new TelemetryEvent(EventPrefix + eventName.ToLower());
+            foreach(var property in properties)
+            {
+                telemetryEvent.Properties.Add(BuildPropertyName(eventName, property.propertyName), property.propertyValue);
+            }
+
+            TryCorrelateProject(unconfiguredProject, telemetryEvent);
+            TelemetryService.DefaultSession.PostEvent(telemetryEvent);
+        }
+
+        /// <summary>
+        /// Hashes personally identifiable information for telemetry consumption.
+        /// </summary>
+        /// <param name="value">Value to hashed.</param>
+        /// <returns>Hashed value.</returns>
+        public string HashValue(string value)
+        {
+            using (var md5 = MD5.Create())
+            {
+                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(value));
+                return BitConverter.ToString(hash);
+            }
+        }
+
+        /// <summary>
+        /// Creates a project asset.
+        /// </summary>
+        /// <param name="unconfiguredProject">Project to create the asset from.</param>
+        /// <returns>Telemetry event representing the project asset.</returns>
+        public async Task CreateProjectAsset(UnconfiguredProject unconfiguredProject)
+        {
+            string assetId = GetProjectId(unconfiguredProject);
```

Comment:
This specific one is creating an asset for a project so this method by nature needs to be tightly coupled.